Implement AnimationPlayer pause() and resume()

madmiraal opened the pull request:
As originally identified in #34125 and elaborated on in https://github.com/godotengine/godot-proposals/issues/287, in `AnimationPlayer`:
> Currently play() method doesn't adhere to single responsibility principle. If user calls play() while there is no animation playing it acts as start_animation() where as when user calls it when animation is already playing it acts as resume_animation(). This dual behaviour is not desired by users as it requires a condition statement in order to use start_animation functionality of play(). It is also inconstant across the engine as for example AudioStreamPlayer method play() method has only singular functionality of start_playing(). This is confusing to understand and breaks intuitiveness of the engine.

Similarly, `AnimationPlayer:stop()` has a parameter that defaults to `true` to either stop and reset the animation or just pause the animation. However, as identified in #27499, although `stop(true)` resets the playback position it doesn't reset the animation.

This PR:
- Implements two new functions in `AnimationPlayer`: `pause()` and `resume()`. `pause()` will pause a currently playing animation, and `resume()` will resume a `paused()` animation.
- Ensures that play() always plays the specified (or current animation) from the beginning regardless of whether or not the requested animation is the same as the current animation
- Ensures that stop() also resets the animation
- Updates the relevant documentation

Closes https://github.com/godotengine/godot-proposals/issues/287
Fixes #27499
Fixes #36279
Supersedes #33733

madmiraal committed:
Implement AnimationPlayer pause() and resume()

Ensures that play() always plays the specified or current animation from
the beginning.

Also fixes stop() not resetting the animation.

akien-mga commented:
We discussed this in a PR review meeting and the general consensus is that it's good, but there are a few things to look further into:

- AnimationPlayer has an `active` property, as well as `is_playing` method, which sound like they might also be related to attempts at pausing playback.  This needs to be looked into some more to clarify.
- AnimatedSprite also has a `playing` property which is now a duplicate for the `pause` and `resume` methods (in #44369).

madmiraal committed:
Remove AnimationPlayer playback_active property

madmiraal committed:
Rename AnimationPlayer::play() to start()

- Removes play_backwards()
- Removes start() from_end parameter. If custom_speed is negative,
  it will automatically play the animation backwards from the end.

madmiraal commented:
> AnimationPlayer has an active property, as well as is_playing method, which sound like they might also be related to attempts at pausing playback. This needs to be looked into some more to clarify.

The `active` property prevents the `AnimationPlayer` from iterating, but it does not pause the playing caches.
The `is_playing()` method returns whether or not an animation has started, but not finished or (with this PR) been paused.


> AnimatedSprite also has a playing property which is now a duplicate for the pause and resume methods (in #44369).

This is one of the issues raised in https://github.com/godotengine/godot-proposals/issues/1994: the need for standardising the naming of `bool` properties and their getters and setters. Specifically, part 5 of the recommendation:
> 5. If the bool property has a natural setter and "unsetter", they can be created as helper functions that map to set_*(true) and set_*(false). For example enabled should have enable() mapped to set_enabled(true) and and disable() mapped to set_enabled(false), paused should have pause mapped to set_paused(true) and resume() mapped to set_paused(false).

madmiraal commented:
Updated to include two new commits:

1. Removes `AnimationPlayer` `playback_active` property and updates the `is_playing()` documentation.
2. Renames `AnimationPlayer::play()` to `start()`, and also removes the `play_backwards()` method and the `from_end` parameter (uses a negative speed to determine this).

madmiraal commented:
Rebased following merge of #44630.

AndreaCatania commented on scene/animation/animation_player.cpp:
```diff
@@ -1325,6 +1327,36 @@ void AnimationPlayer::_animation_changed() {
+void AnimationPlayer::_pause_playing_caches() {
+	for (Set<TrackNodeCache *>::Element *E = playing_caches.front(); E; E = E->next()) {
+		if (E->get()->node && E->get()->audio_playing) {
+			E->get()->node->call("stop");
```
Why relying to `call` instead to store the correct `Node` type and just call `stop`? Or just cast as below.

madmiraal replied:
This is how it's done currently in `_stop_playing_caches()`:
https://github.com/godotengine/godot/blob/9b6c3a871922d63fe5b92288a373789dc7130b4c/scene/animation/animation_player.cpp#L1328-L1332

AndreaCatania commented on scene/animation/animation_player.cpp:
```diff
@@ -1430,17 +1449,17 @@ AnimationPlayer::AnimationMethodCallMode AnimationPlayer::get_method_call_mode()
 	switch (animation_process_mode) {
 		case ANIMATION_PROCESS_PHYSICS:
-			set_physics_process_internal(p_process && active);
+			set_physics_process_internal(p_process);
```
Outside the switch I would put:
```cpp
set_physics_process_internal(false);
set_process_internal(false);
```
So, even if you change the processing mode at runtime, the processing is still correctly set.

madmiraal replied:
Changing this would be outside the scope of this PR. That said, I don't think it's necessary, because setting them to false is already done in the `set_process_callback()`:
https://github.com/godotengine/godot/blob/9b6c3a871922d63fe5b92288a373789dc7130b4c/scene/animation/animation_player.cpp#L1406-L1419

AndreaCatania reviewed (approved):
It looks fine, just few notes and a question (non stopping).

madmiraal commented:
Rebased following merge of #47140.

KoBeWi commented:
The removal of `play()` means that we no longer can continuously play an animation, i.e. you can't
```
func _process(delta):
    animationplayer.play("walk")
```
This is useful sometimes even if it's not really "correct" way to do animations (it reduces code complexity for simple objects).

madmiraal commented:
Updated documentation using `--doctool .`.

reduz commented:
This PR is now redundant given #46191 has been merged, so I am going to close it.

madmiraal commented:
#56645 is a salvage of this PR, because #46191 does not implement godotengine/godot-proposals#287, fix #27499 or fix #36279. Furthermore, pausing (and unpausing) the entire SceneTree is not a convenient way of pausing and resuming an AnimationPlayer's playing animation.
